File: ForwardEuler.py
import numpy as np

def solve_forward_euler(f, t0, t_end, y0, h):
    # f is the differential equation (dy/dt = f(y,t)) (function)
    # t0 is the start time (float)
    # t_end is the final time (float)
    # y0 is the initial value of y (float)
    # h is the step size (must be positive) (float)
    
    if h <= 0:
        raise ValueError("Step size must be positive.")

    N = int((t_end - t0) / h)  # Number of time steps
    t_values = np.linspace(t0, t_end, N+1)
    y_values = np.zeros_like(t_values)
    y_values[0] = y0
    
    for i in range(N):
        t_n = t_values[i]
        y_n = y_values[i]
        
        # Forward Euler update formula
        y_values[i+1] = y_n + h * f(t_n, y_n)
    
    return t_values, y_values

File: NmethodsCodes/ForwardEuler/ForwardEuler.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

// Define the differential equation
double f(double t, double i, double V0, double R, double L) {
    return (V0 - R * i) / L;
}

// Forward Euler function
void Forward_euler(double (*f)(double, double, double, double, double), 
                   double t0, double t_end, double i0, double h, 
                   double V0, double R, double L, double *t_val, double *i_val) {
    
    if (h <= 0) {
        printf("Step size must be positive.\n");
        return;
    }

    int N = (int)((t_end - t0) / h) + 1; // Define number of time steps

    // Initialize values
    t_val[0] = t0;
    i_val[0] = i0;

    // Compute points in loop
    for (int i = 0; i < N - 1 ; i++) {
        t_val[i + 1] = t_val[i] + h; // Update time step
        i_val[i + 1] = i_val[i] + h * f(t_val[i], i_val[i], V0, R, L); // Forward Euler step
    }
}
